Charts analysis

ranger1038 opened the pull request:
implemented initial version of charts analysis;
implemented saving charts properties into database

ranger1038 committed:
implemented initial version of charts analysis;
implemented saving charts properties into database

ranger1038 committed:
analyzing staff was moved into separate files
imports were optimized in comprison.py, job.py, views.py
minor bugs were fixed
survey for data labeling was improved

istillc0de commented on perftracker/models/comparison.py:
```diff
@@ -568,3 +568,32 @@ def init(self, section_ids):
+)
+
+
+class PTComparisonChartTrainDataModel(models.Model):
```
table name will be crazy long

istillc0de commented on perftracker/models/comparison.py:
```diff
@@ -568,3 +568,32 @@ def init(self, section_ids):
+    def pt_validate_json(json_data):
+        if 'data' not in json_data:
+            raise SuspiciousOperation("Chart data is not specified: it must be 'data': <dict>")
+        if type(json_data['data']) is not list:
```
list ? comment says dict

istillc0de commented on perftracker/models/comparison.py:
```diff
@@ -568,3 +568,32 @@ def init(self, section_ids):
+        if type(json_data['data']) is not list:
+            raise SuspiciousOperation("Chart data must be a dict: 'data': <dict> ")
+        if 'incline' not in json_data:
+            raise SuspiciousOperation("Graph incline is not specified: it must be 'incline': <int> ")
```
add newline at the end

istillc0de commented on perftracker/models/comparison.py:
```diff
@@ -568,3 +568,32 @@ def init(self, section_ids):
             g.pt_init_chart_and_table()
+
+
+class PTCmpChartInclineType:
```
consider to move all this to separate file

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
 
+var analyze_was_called = false;
+
+function analyze() {
```
massive non-templated code should go to static.

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
+var analyze_was_called = false;
+
+function analyze() {
+    if (analyze_was_called === true) return;
```
pls:   1) no '=== true '   2) no else

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
+    else analyze_was_called = true;
+
+    var payload = {};
+    Object.keys(sect_data).map(s_id => {
```
map -> forEach

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
+
+    $.ajax({
+        url: api_root + "/analyze",
+        data: JSON.stringify(payload),
```
check if data is NOT double-jsoned

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
+var labeled_data = {};
+
+function label_data(s_id) {
+    labeled_data[s_id] = {data: sect_data[s_id].series['0'].data};
```
what if there is no chart for the section?

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
+
+    var div_to_append = $('#results_' + s_id);
+
+    var selectList = document.createElement('select');
```
may be replace with div_to_append.append("<>...</>")?

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
+function label_data(s_id) {
+    labeled_data[s_id] = {data: sect_data[s_id].series['0'].data};
+
+    var div_to_append = $('#results_' + s_id);
```
consider to add dedicated empty placeholder during template rendering

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -171,6 +172,82 @@ <h4>Job: {{ job }}</h4>
+
+    $.ajax({
+        url: api_root + "/group/{0}/section/{1}/properties".ptFormat(0, s_id),
+        data: JSON.stringify(labeled_data),
```
may be you don't need stringify here

istillc0de commented on perftracker/views.py:
```diff
@@ -170,7 +261,8 @@ def pt_comparison_id_html(request, project_id, cmp_id):
                               'cmp_view': cmp_view,
                               'PTCmpChartType': PTCmpChartType,
-                              'PTCmpTableType': PTCmpTableType
+                              'PTCmpTableType': PTCmpTableType,
```
passing those constants to template seems unwise - they can be hardcoded in static js

istillc0de commented on perftracker/views.py:
```diff
@@ -158,6 +162,93 @@ def pt_comparison_tables_info_json(request, api_ver, project_id, cmp_id, group_i
+    # default value is 1.
+    outlier_coefficient = 1.
+
+    for graph_id, graph_data in data.items():
```
pls move all business logic to separate file, remove dependency from numpy, make sure site can run w/o this functionality

ranger1038 committed:
models import optimization;
added ‘fails’ column for train data table;
changed request method for sending graph data [POST -> PUT];
data validation improved; 
added content-types for Ajax-requests

istillc0de commented on perftracker/static/js/perftracker.js:
```diff
@@ -990,4 +990,81 @@ function pt_isInViewport(elem) {
+
+    var payload = {};
+    Object.keys(sect_data).forEach(s_id => {
+        payload[s_id] = sect_data[s_id].series['0'].data;
```
While I usually recommend to move static code to static file, this is an exception.
The code depends on 'sect_data' variable, and some other variables _not_ defined in perftracker.js, but actually generated by specific view.
This looks strange and misleading and hard to investigate. Either pass that variables as arguments, or embed this code to specific view.

istillc0de replied:
series['0']: 
(1)  0, not '0'
(2) we should probably warn user if there is > 1 jobs in comparison

istillc0de commented on perftracker/static/js/perftracker.js:
```diff
@@ -990,4 +990,81 @@ function pt_isInViewport(elem) {
+
+    $.ajax({
+        url: api_root + "/analyze",
+        data: JSON.stringify(payload),
```
try to add content type and (optionally) skip stringify.
currently data is sent as form

istillc0de commented on perftracker/static/js/perftracker.js:
```diff
@@ -990,4 +990,81 @@ function pt_isInViewport(elem) {
+
+    $.ajax({
+        url: api_root + "/group/{0}/section/{1}/properties".ptFormat(0, s_id),
+        data: JSON.stringify(labeled_data),
```
same issue with content type here, sent as form data

istillc0de commented on perftracker/static/js/perftracker.js:
```diff
@@ -990,4 +990,81 @@ function pt_isInViewport(elem) {
+        timeout: 10000,
+        success: function (response) {
+            Object.keys(sect_data).forEach(s_id => {
+                sect_data[s_id].info = response[s_id];
```
saved for no reason (yet?)

ranger1038 replied:
you're right, I think it may be useful later

istillc0de commented on perftracker/views.py:
```diff
@@ -158,6 +156,48 @@ def pt_comparison_tables_info_json(request, api_ver, project_id, cmp_id, group_i
+    PTTrainDataModel.pt_validate_json(data)
+
+    try:
+        formatted_data = '|'.join([f'{x};{y}' for x, y in data['data']])
```
I suspect there are possibilities for x to be non-number. 
Probably you need to check for that and don't analyze such sections.

istillc0de commented on perftracker/graphs_analysis.py:
```diff
@@ -0,0 +1,57 @@
+    # default value is 1.
+    outlier_coefficient = 1.
+
+    for graph_id, graph_data in data.items():
```
consider to stick to existing terminology - series[_data] 
technically that data is not necessarily 'graph', it can be barchart etc.

ranger1038 committed:
added DEBUG flag on comparison page from django settings;
codestyle changes

istillc0de commented on perftracker/templates/comparison_id.html:
```diff
@@ -161,16 +164,106 @@ <h4>Job: {{ job }}</h4>
+}
+
+function pt_save_graph_properties(s_id) {
+    labeled_data[s_id] = {points: sect_data[s_id].series[0].data};
```
minor: may be points -> serie

istillc0de replied:
important: is it usually important for analysis whether higher is better or lower is better.
We have this information in DB, but not in browser and do not pass here

istillc0de commented on perftracker/views.py:
```diff
@@ -1,47 +1,46 @@
-from perftracker.models.comparison import ComparisonModel, ComparisonSimpleSerializer, ComparisonNestedSerializer, PTComparisonServSideView
-from perftracker.models.regression import RegressionModel, RegressionSimpleSerializer, RegressionNestedSerializer, PTRegressionServSideView
+from perftracker.forms import PTCmpDialogForm, PTHwFarmNodeLockForm, PTJobUploadForm
+from perftracker.series_analysis import pt_comparison_series_analyze
```
consider to move that import right into 
def pt_comparison_analyze_json() ...

istillc0de commented on perftracker_django/urls.py:
```diff
@@ -38,6 +38,10 @@
         views.pt_comparison_tables_info_json),
     url(r'^api/v(?P<api_ver>\d+.\d+)/(?P<project_id>\d+)/comparison/(?P<cmp_id>\d+)/tables$',
         views.pt_comparison_tables_info_json),
+    url(r'^api/v(?P<api_ver>\d+.\d+)/(?P<project_id>\d+)/comparison/(?P<cmp_id>\d+)/analyze$',
```
may be it is time to stop forming/using wrong URIs
1) (minor) api_ver can handled not per uri but some common root (I bet Django has support for that)
2) don't pass/don't parse unused parameters ( e.g. project_id) - only those that matter. For pt_comparison_section_properties_save none matters, but may be it is reasonable to keep cmp_id + section_id just fore reference.

istillc0de commented on perftracker/views.py:
```diff
@@ -38,7 +36,10 @@
+from perftracker.models.train_data import TrainDataModel, CHART_FUNCTION_TYPES, \
+    CHART_OUTLIERS, CHART_OSCILLATION, CHART_ANOMALY
 from perftracker.rest import pt_rest_ok, pt_rest_err, pt_rest_not_found, pt_rest_method_not_allowed
+from perftracker_django.settings import DEBUG
```
don't think DEBUG is the right one.
my proposal was to add new flag that enables your logic

ranger1038 committed:
minor changes:  
  - renaming [points -> serie] 
  - design changes 
  - added flag for analysis hiding
  - added less_better field

ranger1038 committed:
changed less_better field handling

ranger1038 committed:
added migration
